crates/node-builder/src/builder.rs

```diff
@@ -0,0 +1,786 @@
+            beacon_engine_handle.event_listener().map(Into::into),
+            pipeline_events.map(Into::into),
+            if config.debug.tip.is_none() {
+                Either::Left(
+                    ConsensusLayerHealthEvents::new(Box::new(blockchain_db.clone()))
+                        .map(Into::into),
+                )
+            } else {
+                Either::Right(stream::empty())
+            },
+            pruner_events.map(Into::into)
+        );
+        executor.spawn_critical(
+            "events task",
+            reth_node_core::events::node::handle_events(
+                Some(network.clone()),
+                Some(head.number),
+                events,
+                database.clone(),
+            ),
+        );
+
+        let engine_api = EngineApi::new(
+            blockchain_db.clone(),
+            config.chain.clone(),
+            beacon_engine_handle,
+            payload_builder.into(),
+            Box::new(executor.clone()),
+        );
+        info!(target: "reth::cli", "Engine API handler initialized");
+
+        // extract the jwt secret from the args if possible
+        let default_jwt_path = data_dir.jwt_path();
+        let jwt_secret = config.rpc.auth_jwt_secret(default_jwt_path)?;
+
+        // adjust rpc port numbers based on instance number
+        config.adjust_instance_ports();
+
+        // Start RPC servers
+
+        let (rpc_server_handles, rpc_registry) = crate::rpc::launch_rpc_servers(
+            node_components.clone(),
+            engine_api,
+            &config,
+            jwt_secret,
+            rpc,
+        )
+        .await?;
+
+        // Run consensus engine to completion
+        let (tx, rx) = oneshot::channel();
+        info!(target: "reth::cli", "Starting consensus engine");
+        executor.spawn_critical_blocking("consensus engine", async move {
+            let res = beacon_consensus_engine.await;
+            let _ = tx.send(res);
+        });
+
+        let FullNodeComponentsAdapter {
+            evm_config,
+            pool,
+            network,
+            provider,
+            payload_builder,
+            executor,
+        } = node_components;
+
+        let full_node = FullNode {
+            evm_config,
+            pool,
+            network,
+            provider,
+            payload_builder,
+            executor,
+            rpc_server_handles,
+            rpc_registry,
+            config,
+            data_dir,
+        };
+        // Notify on node started
+        on_node_started.on_event(full_node.clone())?;
+
+        let handle = NodeHandle {
+            node_exit_future: NodeExitFuture::new(rx, full_node.config.debug.terminate),
+            node: full_node,
+        };
+
+        Ok(handle)
+    }
+
+    /// Check that the builder can be launched
+    ///
+    /// This is useful when writing tests to ensure that the builder is configured correctly.
+    pub fn check_launch(self) -> Self {
```

Comment:
should this be #[cfg(test)] gated?

Reply:
this is useful when writing e2e tests. this can stay